### PR TITLE
TW-1106: Storage analytics

### PR DESCRIPTION
**Pull request by alex-tsx**
https://madfish.atlassian.net/browse/TW-1106

### CODE REVIEW

File: src/store/index.ts [hunk 1]
```diff
@@ -42,3 +42,5 @@ export const { store, persistor } = createStore(
   farmsEpics,
   savingsEpics
 );
+
+export const dispatch = store.dispatch.bind(store);
```
Comment (lendihop) [inline src/store/index.ts:46]: what's the problem with useDispatch?

File: src/declarations.d.ts [hunk 2]
```diff
@@ -1,5 +1,7 @@
 type nullish = null | undefined;
 
+type EmptyFn = () => void;
```
Comment (lendihop) [inline src/declarations.d.ts:3]: this fn is already exist in general.ts

File: src/hooks/root-hooks/use-storage-analytics.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+const getAsyncStorageUsageDetails = async () => {
+  const keys = await AsyncStorage.getAllKeys();
+  const values = await AsyncStorage.multiGet(keys);
+
+  const sizesByKey: Record<string, number> = {};
+  let totalValuesSize = 0;
+
+  for (const [key, value] of values) {
+    const size = value ? calculateStringSizeInBytes(value) : 0;
+    totalValuesSize += size;
```
Comment (lendihop) [inline src/hooks/root-hooks/use-storage-analytics.ts:33]: mb we need some flag like `isCriticalSize = totalValuesSize > some value`?

File: src/hooks/root-hooks/use-storage-analytics.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  const measureAndSend = useCallback(
+    () =>
+      getAsyncStorageUsageDetails().then(asyncStorage =>
+        trackEvent('STORAGES_STATE', AnalyticsEventCategory.General, { asyncStorage })
+      ),
+    [trackEvent]
+  );
+
+  // Delaying to not throttle initial app load
+  useTimeout(measureAndSend, 10000);
```
Comment (lendihop) [inline src/hooks/root-hooks/use-storage-analytics.ts:21]: why 10s? mb there is a device where initial load would take longer.. and if there is a problem with throttle mb we need the same timeout in `usePushNotifications`?

File: src/hooks/root-hooks/use-storage-analytics.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+const getAsyncStorageUsageDetails = async () => {
+  const keys = await AsyncStorage.getAllKeys();
+
+  const records = await Promise.all(
+    keys.map(key =>
+      AsyncStorage.getItem(key).then(
+        val => [key, val] as const,
+        error => {
+          console.error(error);
```
Comment (lourenc) [inline src/hooks/root-hooks/use-storage-analytics.ts:28]: Could you, please, clarify one thing? AFAIU the issue we have been trying to track is related to AsyncStorage not being able to `setItem`,  but not `getItem`. So if the key is "oversized" as you call it, we won't be able to set it at all and most likely won't catch the error here 🤔

**Review by lourenc — APPROVED** [review]

File: src/hooks/root-hooks/use-storage-analytics.ts (hunk 5 above)
Comment (lourenc) [reply src/hooks/root-hooks/use-storage-analytics.ts:28]: Ok! Thank you for clarification

**Review by lendihop — APPROVED** [review]

**Review by NovakVlad2424 — APPROVED** [review]
ios done

**Review by IrynaKhyzhynska — APPROVED** [review]
android done

**Pull request merged**